common/lib/common-definitions/src/logger.ts

```diff
@@ -45,12 +45,166 @@ export interface ITelemetryBaseEvent extends ITelemetryProperties {
     eventName: string;
 }
 
+export type TelemetryErrorCategory = "verbose" | "information" | "warning" | "error" | "critical";
+
+/**
+ * General-use event. These are statically typed events that clients can understand and act on.
+ */
+export interface ITelemetryGenEventBase extends ITelemetryBaseEvent {
+    /*
+     * Flag indicating this is general use event.
+     * Note: It may seem redundant if we end up emitting general use-events separately.
+     * However, it still may be useful if all (dev and general-use) telemetry is routed to a single table.
+     */
+    genUse: true;
+    /*
+     * Subtype (Further description in sublasses).
+     */
+    type: "api" | "service" | "event" | "error";
+    /*
+     * Package this event was emitted from.
+     * It may be useful for consumers to filter out specific packages they are working with.
+     */
+    packageName: string;
+    /*
+     * Class name where this event was emitted from.
+     * It may be useful for consumers to filter out specific classes they are working with.
+     */
+    className: string;
+    /*
+     * Fluid document ID (when applicable). This could be ID, resolved URL etc.
+     */
+    docId?: string;
+    /*
+     * Id of a client who generated this telemetry event (when applicable).
+     */
+    clientId?: string;
+}
+
+/**
+ * General-use API event. Generated to log a request received by API.
+ */
+export interface ITelemetryGenApiEvent extends ITelemetryGenEventBase {
+    type: "api";
+    /*
+     * API name that generated this telemetry.
+     */
+    apiName: string;
+    /*
+     * Duration of the call.
+     */
+    duration: number;
+    /*
+     * Identifier of a request call instance.
+     * Used for correlation (TODO) between request and other telemetry items.
+     */
+    id: string;
+    /*
+     * Flag indicating if call suceeded or failed.
+     */
+    success: boolean;
+    /*
+     * Optional response status.
+     */
+    status?: string;
+    /*
+     * JSON stringifed bag od details on the event.
+     */
+    details?: string;
+}
+
+/**
+ * General-use Error event. Typically represents an exception that causes an operation to fail.
+ */
+export interface ITelemetryGenErrorEvent extends ITelemetryGenEventBase {
+    type: "error";
+    /*
+     * Used for exceptions grouping (ex exception type + function).
+     */
+    errorCode: string;
+    /*
+     * Optional message.
+     */
+    message?: string;
+    /*
+     * Severity level.
+     */
+    severityLevel: TelemetryErrorCategory;
+    /*
+     * JSON stringifed stack trace
+     */
+    stackTrace?: string;
+    /*
+     * JSON stringifed details on condition(s) leading to this error
+     */
+    details?: string;
```

Comment:
What is the story for handling PII or user data of any sorts?
I.e. if we assume this kind of API will be used by 1st party, we need to know if data may contain any user data. We use tagging in logging. For example, DDS IDs, or packages are tagged with TelemetryDataTag.CodeArtifact, indicating that that data may contained user data.  I believe same applies to stack property.

TelemetryDataTag.CodeArtifact means - we do not know for sure, the app that does actual logging need to determine if it's safe (for example, it's safe for Loop as they do not use user data to name DDSs, and do not host any 3rd party code, so stack does not expose any concepts that we should not collect). But same maybe not true for other apps.

Stack is easy (as it's concrete concept), but other it's not clear how to deal with let's say `details` from this angle

Reply:
Yes, this is an interesting topic.

Maybe let me describe how I see this change, and problem segmentation. Some answers here may likely (and hopefully:) answer some of the other questions above. 

1. We are extending existing ITelemetryGenericEvent and ITelemetryBaseLogger structure to offer set of telemetry that is usable/actionable to general clients.
2. In the very conservative case,(1) may end up being just another layer of categorization, (see other comment on categories), so that general user knows the intended purpose of various categories and how to act on the telemetry data.
4. In the more optimistic case this telemetry offers additional structured information around concepts that we know will stick around for a longer time (ex. clientIds, but based on your earlier comment that may go down to the trash bucket). Maybe we can be opinionated about specific categories (ex. errors, perf), and less opinionated about others as a general guidelines.
5. We will pay the “change” cost here (back compat etc). We can limit that cost by keeping the structure increment smaller. Of course, behavior change (when telemetry gets emitted) is harder to manage. But so is any other behavior change within FF. Dev has to conclude is it breaking (customer needs to spend effort to adapt) or not.Also, telemetry (and its change in behavior) may play a helpful role here in pointing towards some unintentional change that has occurred in the system.
6. For fields such as “detail” I bucketed them into “we still don’t need to provide any guarantees” and “anyways we are emitting them today”. We can informally describe how we go about tagging/PII, but don’t need to offer any further guarantees.

Reply:
I'm not sure I follow. You are saying that the goal is not to have parallel / duplicate events. This means any existing logging statement may in future use this system. How do I express in this system that some property (that is being hidden in details) may contain PII and that Bohemia may need to filter it out before pushing to Aria?

We have pretty clear guidance today - see ITaggedTelemetryPropertyType